### PR TITLE
[oneTBB] Add spec changes for std::invoke feature

### PR DESCRIPTION
**Pull request by aepanchi**

### CODE REVIEW

File: source/elements/oneTBB/source/algorithms/functions/parallel_for_func.rst [hunk 1]
```diff
@@ -56,19 +56,27 @@ A ``partitioner`` type may be one of the following entities:
 Requirements:
 
 * The ``Range`` type must meet the :doc:`Range requirements <../../named_requirements/algorithms/range>`.
-* The ``Body`` type must meet the :doc:`ParallelForBody requirements <../../named_requirements/algorithms/par_for_body>`.
 * The ``Index`` type must meet the :doc:`ParallelForIndex requirements <../../named_requirements/algorithms/par_for_index>`.
+
+Until C++17:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/algorithms/functions/parallel_for_func.rst:61]:
```suggestion
Before C++17:
```

File: source/elements/oneTBB/source/named_requirements/algorithms/cpp17_par_scan_func.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+=====================
+**[req.cpp17_parallel_scan_func]**
+
+A type `Func` satisfies `Cpp17ParallelScanFunc` if it meets
+the `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const Range&, const Value&, bool`` and return the type ``Value``
+where ``Range`` and ``Value`` are the same as passed to ``parallel_scan``.
+
+Considering ``func`` to be an object of type ``const Func``, ``range`` to be an object of type ``Range``, ``sum`` be an object of type ``Value``
+and ``is_final`` be an object of type ``bool``,
+applying ``std::invoke(func, range, sum, is_final)`` starting with ``sum``, computes the summary and, for ``is_final == true``,
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/algorithms/cpp17_par_scan_func.rst:16]:
```suggestion
applying ``std::invoke(func, range, sum, is_final)``, starting with a ``sum``, computes the summary and, for ``is_final == true``,
```

File: source/elements/oneTBB/source/named_requirements/algorithms/cpp17_par_scan_func.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+the scan result for range ``range``. Returns the computed summary.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/algorithms/cpp17_par_scan_func.rst:17]:
```suggestion
the scan result for the ``range`` range. Returns the computed summary.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 4]
```diff
@@ -0,0 +1,23 @@
+.. SPDX-FileCopyrightText: 2019-2023 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+==================
+Cpp17AsyncNodeBody
+==================
+**[req.cpp17_async_node_body]**
+
+A type `Body` satisfies `Cpp17AsyncNodeBody` if it meets the following requirements:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:10]:
```suggestion
The ``Body`` type  satisfies `Cpp17AsyncNodeBody` if it meets the following requirements:
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+* The `CopyConstructible` requirements from [copyconstructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:12]:
```suggestion
* The `CopyConstructible` requirements from the [copyconstructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+* The `Destructible` requirements from [destructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:13]:
```suggestion
* The `Destructible` requirements from the  [destructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+* The `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const Input&, GatewayType&`` and return the type ``void``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:14]:
```suggestion
* The `Callable` requirements from the [func.wrap.func] section of the ISO C++ Standard for the ``const Input&, GatewayType&`` arguments types and returns the ``void`` type. 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  where ``Input`` is the corresponding template argument of the ``async_node``. The ``GatewayType`` is the same as the ``gateway_type`` member type
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:15]:
```suggestion
Where ``Input`` is the corresponding template argument of the ``async_node``. The ``GatewayType`` is the same as the ``gateway_type`` member type
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  of the ``async_type``.
+
+Considering ``body`` to be an object of type ``Body``, ``v`` to be an object of type ``const Value&`` and ``gateway`` be an object of type ``GatewayType&``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:18]:
```suggestion
Considering:
* The ``body`` to be an object of the ``Body`` type 
* The ``v`` to be an object of the ``const Value&`` type
* The ``gateway`` to be an object of the ``GatewayType&`` type 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+applying ``std::invoke(body, v, gateway)`` submits the value ``v`` to the external activity.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:19]:
```suggestion
applying ``std::invoke(body, v, gateway)`` submits the ``v`` value to the external activity.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+The :doc:`gateway interface <gateway_type>` allows the external activity to communicate
+with the enclosing flow graph.
+
+**_NOTE:_**  It allows to pass pointer-to-function of the class ``Input`` as a body of ``async_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_asyn_node_body.rst:23]:
```suggestion
**_NOTE:_**  You can pass a pointer-to-function of the class  as a body of ``async_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 12]
```diff
@@ -0,0 +1,20 @@
+.. SPDX-FileCopyrightText: 2019-2023 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=====================
+Cpp17FunctionNodeBody
+=====================
+**[req.cpp17_function_node_body]**
+
+A type `Body` satisfies `Cpp17FunctionNodeBody` if it meets the following requirements:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:10]:
```suggestion
The `Body` type satisfies `Cpp17FunctionNodeBody` if it meets the following requirements:
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+* The `CopyConstructible` requirements from [copyconstructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:12]:
```suggestion
* The `CopyConstructible` requirements from the [copyconstructible] section of the ISO C++ Standard. 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+* The `Destructible` requirements from [destructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:13]:
```suggestion
* The `Destructible` requirements from the [destructible] section of the ISO C++ Standard. 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+* The `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const Input&`` and return the type ``Output``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:14]:
```suggestion
* The `Callable` requirements from the [func.wrap.func] section of the ISO C++ Standard  for the ``const Input&`` arguments types and returns the ``Output`` type. 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  where ``Input`` and ``Output`` are corresponding template arguments of the ``function_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:15]:
```suggestion
Where ``Input`` and ``Output`` are the corresponding template arguments of the ``function_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+Considering ``body`` to be an object of type ``Body`` and ``input`` to be an object of type ``const Input&``,
+applying ``std::invoke(body, input)`` performs operation on the value ``input``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:18]:
```suggestion
applying ``std::invoke(body, input)`` performs operation on the ``input`` value.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+**_NOTE:_**  It allows to pass pointer-to-function or pointer-to-member of the class ``Input`` as a body of ``function_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_function_node_body.rst:20]:
```suggestion
**_NOTE:_** You can pass a pointer-to-function or a pointer-to-member of the ``Input`` class as a body of ``function_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 19]
```diff
@@ -0,0 +1,21 @@
+.. SPDX-FileCopyrightText: 2019-2023 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+===========================
+Cpp17JoinNodeFunctionObject
+===========================
+**[req.cpp17_join_node_function_object]**
+
+A type `FunctionObject` satisfies `Cpp17JoinNodeFunctionObject` if it meets the following requirements:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:10]:
```suggestion
The ``FunctionObject`` type satisfies `Cpp17JoinNodeFunctionObject` if it meets the following requirements:
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+* The `CopyConstructible` requirements from [copyconstructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:12]:
```suggestion
* The `CopyConstructible` requirements from the [copyconstructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+* The `Destructible` requirements from [destructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:13]:
```suggestion
* The `Destructible` requirements from the [destructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+* The `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const Input&`` and return the type ``Key``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:14]:
```suggestion
* The `Callable` requirements from the [func.wrap.func] section of the ISO C++ Standard for the ``const Input&`` arguments types and returns the ``Key`` type.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+  where ``Key`` is the same as the corresponding template argument of ``join_node`` and the ``Input`` is the corresponding
+  element of the ``OutputTuple`` template argument of the ``join_node``.
+
+Considering ``f`` to be an object of type ``FunctionObject`` and ``input`` to be an object of type ``const Input&``,
+applying ``std::invoke(f, input)`` returns the key to be used for hashing ``input``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:19]:
```suggestion
applying ``std::invoke(f, input)`` returns the key to be used for hashing the ``input``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+
+**_NOTE:_**  It allows to pass pointer-to-function or pointer-to-member of the class ``Input`` as a body of ``join_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_join_node_func_obj.rst:21]:
```suggestion
**_NOTE:_** You can pass a pointer-to-function or a pointer-to-member of the ``Input`` class as a body of ``join_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 25]
```diff
@@ -0,0 +1,22 @@
+.. SPDX-FileCopyrightText: 2019-2023 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+==========================
+Cpp17MultifunctionNodeBody
+==========================
+**[req.cpp17_multifunction_node_body]**
+
+A type `Body` satisfies `Cpp17MultifunctionNodeBody` if it meets the following requirements:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:10]:
```suggestion
The ``Body`` type satisfies `Cpp17MultifunctionNodeBody` if it meets the following requirements:
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+
+* The `CopyConstructible` requirements from [copyconstructible] ISO C++ Standard section
+* The `Destructible` requirements from [destructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:13]:
```suggestion
* The `Destructible` requirements from the [destructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+* The `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const Input&, OutputPortsType&`` and return the type ``void``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:14]:
```suggestion
* The `Callable` requirements from the [func.wrap.func] section of the ISO C++ Standard  for the ``const Input&, OutputPortsType&`` arguments types, and returns the ``void`` type.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  where ``Input`` is the corresponding template argument of the ``multifunction_node``. The ``OutputPortsType`` is the same as the ``output_ports_type`` member type
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:15]:
```suggestion
Where the ``Input`` is the corresponding template argument of the ``multifunction_node``. The ``OutputPortsType`` is the same as the ``output_ports_type`` member type
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+  of the ``multifunction_node``.
+
+Considering ``body`` to be an object of type ``Body``, ``input`` to be an object of type ``const Input&`` and ``output_ports``` to be an object of type ``OutputPortsType&``,
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:18]:
```suggestion
Considering:

 * The ``body`` to be an object of the ``Body`` type
 * The ``input`` to be an object of the ``const Input&`` type
 * The ``output_ports``` to be an object of the ``OutputPortsType&`` type
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+applying ``std::invoke(body, input, output_ports)`` performs operation on ``input``. May call ``try_put()`` on zero or more of the output ports in ``output_ports``.
+May call ``try_put()`` on any output port multiple times.
+
+**_NOTE:_**  It allows to pass pointer-to-function of the class ``Input`` as a body of ``multifunction_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_multifunction_node_body.rst:22]:
```suggestion
**_NOTE:_** You can pass a pointer-to-function of the ``Input`` class as a body of ``multifunction_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 31]
```diff
@@ -0,0 +1,20 @@
+.. SPDX-FileCopyrightText: 2019-2023 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+==============
+Cpp17Sequencer
+==============
+**[req.cpp17_sequencer]**
+
+A type `Sequencer` satisfies `Cpp17FunctionNodeBody` if it meets the following requirements:
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:10]:
```suggestion
The ``Sequencer`` type satisfies `Cpp17FunctionNodeBody` if it meets the following requirements:
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+
+* The `CopyConstructible` requirements from [copyconstructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:12]:
```suggestion
* The `CopyConstructible` requirements from the [copyconstructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+* The `Destructible` requirements from [destructible] ISO C++ Standard section
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:13]:
```suggestion
* The `Destructible` requirements from the [destructible] section of the ISO C++ Standard.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+* The `Callable` requirements from [func.wrap.func] ISO C++ Standard section for arguments types ``const T&`` and return the type ``size_t``
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:14]:
```suggestion
* The `Callable` requirements from the [func.wrap.func] section of the ISO C++ Standard for the ``const T&`` arguments types, and returns the ``size_t`` type. 
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 35]
```diff
@@ -0,0 +1,20 @@
...
+  where ``T`` is the corresponding template argument of the ``sequencer_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:15]:
```suggestion
Where ``T`` is the corresponding template argument of the ``sequencer_node``.
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+
+Considering ``s`` to be an object of type ``Sequencer`` and ``v`` to be an object of type ``const T&``,
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:17]:
```suggestion
Considering ``s`` to be an object of the ``Sequencer`` type and ``v`` to be an object of the ``const T&`` type,
```

File: source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst [hunk 37]
```diff
@@ -0,0 +1,20 @@
...
+applying ``std::invoke(s, v)`` returns the sequence number for the provided message ``v``.
+
+**_NOTE:_**  It allows to pass pointer-to-function or pointer-to-member of the class ``T`` as a body of ``sequencer_node``.
```
Comment (aepanchi) [inline source/elements/oneTBB/source/named_requirements/flow_graph/cpp17_sequencer.rst:20]:
```suggestion
**_NOTE:_**  You can pass a pointer-to-function or a pointer-to-member of the ``T`` class as a body of ``sequencer_node``.
```

**Comment by aepanchi** [comment]
You should also include all of the new files to the toctree.

**Comment by rscohn2** [comment]
Ready to merge?

**Comment by akukanov** [comment]
Closing this PR in favor of https://github.com/oneapi-src/oneAPI-spec/pull/510

**Pull request closed without merging**